styleguide_example/blog_examples/google_login_server_flow/sdk/services.py

```diff
@@ -0,0 +1,91 @@
+import requests
+from attr import define
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured, ValidationError
+from django.urls import reverse_lazy
+
+
+@define
+class GoogleSDKLoginCredentials:
+    secret_file: str
+
+
+class GoogleSDKLoginFlowService:
+    API_URI = reverse_lazy("api:google-oauth2:login-sdk:callback-sdk")
+
+    SCOPES = [
+        "https://www.googleapis.com/auth/userinfo.email",
+        "https://www.googleapis.com/auth/userinfo.profile",
+        "openid",
+    ]
+    GOOGLE_USER_INFO_URL = "https://www.googleapis.com/oauth2/v3/userinfo"
+
+    def __init__(self):
+        self._credentials = google_sdk_login_get_credentials()
+
+    @staticmethod
+    def _jwt_decode_token(token: str) -> Dict[str, Any]:
+        decoded_data = jwt.decode(jwt=token, options={"verify_signature": False})
+        return decoded_data
+
+    def _get_redirect_uri(self):
+        domain = settings.BASE_BACKEND_URL
+        api_uri = self.API_URI
+        redirect_uri = f"{domain}{api_uri}"
+        return redirect_uri
+
+    def get_authorization_url(self):
+        redirect_uri = self._get_redirect_uri()
+        # Reference: https://developers.google.com/identity/protocols/oauth2/web-server#creatingclient
+        google_oauth_flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+            self._credentials.secret_file, scopes=self.SCOPES
+        )
+        google_oauth_flow.redirect_uri = redirect_uri
+
+        authorization_url, state = google_oauth_flow.authorization_url(
+            access_type="offline",
+            include_granted_scopes="true",
+            prompt="select_account",
+        )
+        return authorization_url, state
+
+    def get_access_credentials(self, *, code: str, state: str):
+        redirect_uri = self._get_redirect_uri()
+        flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+            self._credentials.secret_file, scopes=self.SCOPES, state=state
+        )
+        flow.redirect_uri = redirect_uri
+        access_credentials_payload = flow.fetch_token(code=code)
+        if not access_credentials_payload:
+            raise ValidationError("Failed to obtain access credentials from Google")
+
+        return access_credentials_payload
+
+    def get_user_info(self, *, access_token: str) -> Dict[str, Any]:
+        # Reference: https://developers.google.com/identity/protocols/oauth2/web-server#callinganapi
+        response = requests.get(self.GOOGLE_USER_INFO_URL, params={"access_token": access_token})
+
+        if not response.ok:
+            raise ValidationError("Failed to obtain user info from Google.")
+
+        return response.json()
+
+    def get_token_decoded(self, *, token: str) -> Dict[str, Any]:
+        decoded_token = self._jwt_decode_token(token)
+        return decoded_token
+
+
+def google_sdk_login_get_credentials():
+    secret_file = "google_client_secret.json"
+
+    if not os.path.exists(secret_file):
```

Comment:
This would be problematic for deployments, since it'll require the actual file to be part of the project.

We'd rather want to read it as a string from an env and load it via string, for the Google SDK :+1: